src/item.cpp

```diff
@@ -12354,21 +12205,16 @@ const item_category &item::get_category_shallow() const
     return type->category_force.is_valid() ? type->category_force.obj() : null_category;
 }
 
-const item_category &item::get_category_of_contents() const
+const item_category &item::get_category_of_contents( int depth, int maxdepth ) const
 {
-    if( type->category_force == item_category_container ) {
-        const std::list<const item *> items = contents.all_items_top( pocket_type::CONTAINER );
-        if( !items.empty() ) {
-            const item_category &category_of_first_item = items.front()->get_category_of_contents();
-            if( items.size() == 1 ) {
-                return category_of_first_item;
-            }
-            const auto has_same_category = [&category_of_first_item]( const item * i ) {
-                return i->get_category_of_contents() == category_of_first_item;
-            };
-            if( std::all_of( ++items.begin(), items.end(), has_same_category ) ) {
-                return category_of_first_item;
-            }
+    if( depth++ < maxdepth && type->category_force == item_category_container ) {
+        if( cached_category.timestamp == calendar::turn ) {
+            return cached_category.id.obj();
+        }
+        if( item::aggregate_t const aggi = aggregated_contents( depth, maxdepth ); aggi ) {
```

Comment:
I prefer not having side effects like that buried in oneline conditional where the actual use comes later.
```suggestion
    if( depth < maxdepth && type->category_force == item_category_container ) {
        if( cached_category.timestamp == calendar::turn ) {
            return cached_category.id.obj();
        }
        ++depth;
        if( item::aggregate_t const aggi = aggregated_contents( depth, maxdepth ); aggi ) {
```